util/tlgen/item.py

```diff
@@ -2,58 +2,45 @@
 # Licensed under the Apache License, Version 2.0, see LICENSE for details.
 # SPDX-License-Identifier: Apache-2.0
 
-from enum import Enum
+"""Classes to represent the crossbar's dataflow graph."""
+
+from typing import List, Tuple
 
 
 class Edge:
-    """Edge class contains the connection from a node to a node.
+    """A connection between two nodes.
 
-    a Node can be a host port, output of async_fifo, port in a socket,
+    A node can be a host port, output of async_fifo, port in a socket,
     or a device port.
     """
-    def __init__(self, us, ds):
+
+    def __init__(self, us: 'Node', ds: 'Node'):
+        """Create an edge between us and ds."""
         self.us = us
         self.ds = ds
 
-    def __repr__(self):
-        return "U(%s) D(%s)" % (self.us.name, self.ds.name)
-
-
-# Edges = List[Edge]
-# Clocks = List[str]  # If length is more than one, should be exactly two
-
-# [UpstreamClock, DownstreamClock]
-
-
-class NodeType(Enum):
-    HOST = 1
-    DEVICE = 2
-    ASYNC_FIFO = 3
-    SOCKET_1N = 4
-    SOCKET_M1 = 5
-
 
 class Node:
-    """Node class is a port that communicates from/to other Node or TL-UL
-    input/output.
-    """
+    """Base class representing a node in the data graph."""
+
+    name: str = ""
 
-    name = ""  # name: str
-    # node_type: NodeType
-    clocks = []  # Clocks  # clock domains of the node
-    resets = []  # Resets  # resets of the node
+    clocks: List[str] = []  # clock domains of the node. Should have at most 2.
+    resets: List[str] = []  # resets of the node
     # e.g. async_fifo in : clk_core , out : clk_main
 
     # If NodeType is Socket out from 1:N then address steering is used
     # But this value is also propagated up to a Host from multiple Devices
-    # Device Node should have address_from, address_to
-    # address_from = 0  #: int
-    # address_to = 0  #: int
-    addr_range = []
-
-    us = []  # Edges  # Number of Ports depends on the NodeType
+    # A device Node should have address_from, address_to.
+    #
+    # Other nodes have this fields set to -1.
+    address_from: int = -1
+    address_to: int = -1
+    addr_range: List[Tuple[int, int]] = []
+
+    us: 'List[Edge]' = []  # Number of Ports depends on the NodeType
```

Comment:
I don't think you need the forward-declaration quoting here?

Reply:
Hmm, `mypy` actually complains without it! :-)
```
rjs@sole:~/work/opentitan/util/tlgen$ mypy --strict *.py
item.py:41: error: Need type annotation for "us" (hint: "us: List[<type>] = ...")
Found 1 error in 1 file (checked 9 source files)
```

I think the point is that it can't cope with "declare a variable; fill it in later" without type hints. So `foo = 123` doesn't need to be `foo: int = 123`. But `foo = []` isn't enough for MyPy because it doesn't know what type lives in the list.